File: java/Ice/inheritance/server/src/main/java/com/example/ice/filesystem/server/MDirectory.java
// Copyright (c) ZeroC, Inc.

package com.example.ice.inheritance.server;

import com.example.inheritance.Directory;
import com.example.inheritance.NodePrx;

import com.zeroc.Ice.Current;

import java.util.ArrayList;

/**
 * Provides an in-memory implementation of the Slice interface Directory.
 */
class MDirectory extends MNode implements Directory {
    private ArrayList<NodePrx> _contents = new ArrayList<>();

    /**
     * Constructs an {@code MDirectory}.
     *
     * @param name the name of this directory
     */
    MDirectory(String name) {
        super(name);
    }

    @Override
    public NodePrx[] list(Current current) {
        return _contents.toArray(new NodePrx[0]);
    }

    /**
     * Adds a node to this directory
     *
     * @param child the node proxy to add
     */
    void addChild(NodePrx child) {
        _contents.add(child);
    }
}

File: csharp/Ice/Inheritance/Server/MDirectory.cs
// Copyright (c) ZeroC, Inc.

namespace Server;

/// <summary>Provides an in-memory implementation of the Slice interface Directory.</summary>
internal class MDirectory : Inheritance.DirectoryDisp_
{
    private readonly MNode _node; // reuse the MNode implementation
    private readonly List<Inheritance.NodePrx> _contents = [];

    /// <inheritdoc/>
    // Implements Node.Name by forwarding to _node.
    public override string Name(Ice.Current current) => _node.Name(current);

    /// <inheritdoc/>
    public override Inheritance.NodePrx[] List(Ice.Current current) => _contents.ToArray();

    /// <summary>Constructs an MDirectory.</summary>
    /// <param name="name">The name of this directory.</param>
    internal MDirectory(string name) => _node = new MNode(name);

    /// <summary>Adds a node to this directory.</summary>
    /// <param name="child">The node proxy to add.</param>
    internal void AddChild(Inheritance.NodePrx child) => _contents.Add(child);
}
